File: 2022/day05/python/solution.py
#!/usr/bin/env python3
import os
import re
import copy

def parse_input(filename):
    with open(filename) as f:
        content = f.read()

    parts = content.split('\n\n')
    stack_lines = parts[0].split('\n')
    move_lines = parts[1].strip().split('\n')

    # Find number of stacks from the last line (the numbers)
    num_stacks = len(stack_lines[-1].split())

    # Parse stacks (bottom-up, excluding the number line)
    stacks = [[] for _ in range(num_stacks)]
    for line in stack_lines[:-1]:
        for i in range(num_stacks):
            pos = 1 + i * 4  # Position of crate letter
            if pos < len(line) and line[pos] != ' ':
                stacks[i].append(line[pos])

    # Reverse so bottom is at index 0
    for stack in stacks:
        stack.reverse()

    # Parse moves
    moves = []
    for line in move_lines:
        match = re.match(r'move (\d+) from (\d+) to (\d+)', line)
        if match:
            count = int(match.group(1))
            from_stack = int(match.group(2)) - 1  # 0-indexed
            to_stack = int(match.group(3)) - 1
            moves.append((count, from_stack, to_stack))

    return stacks, moves

def part1(stacks, moves):
    stacks = copy.deepcopy(stacks)
    for count, from_stack, to_stack in moves:
        for _ in range(count):
            crate = stacks[from_stack].pop()
            stacks[to_stack].append(crate)
    return ''.join(stack[-1] for stack in stacks if stack)

def part2(stacks, moves):
    stacks = copy.deepcopy(stacks)
    for count, from_stack, to_stack in moves:
        # Move multiple crates at once (preserve order)
        crates = stacks[from_stack][-count:]
        stacks[from_stack] = stacks[from_stack][:-count]
        stacks[to_stack].extend(crates)
    return ''.join(stack[-1] for stack in stacks if stack)

def main():
    script_dir = os.path.dirname(os.path.abspath(__file__))
    input_file = os.path.join(script_dir, '..', 'input.txt')

    stacks, moves = parse_input(input_file)

    print('Part 1:', part1(stacks, moves))
    print('Part 2:', part2(stacks, moves))

if __name__ == '__main__':
    main()

File: 2022/day05/ruby/solution.rb
#!/usr/bin/env ruby

def parse_input(filename)
  content = File.read(filename)
  parts = content.split("\n\n")
  stack_lines = parts[0].split("\n")
  move_lines = parts[1].strip.split("\n")

  # Find number of stacks from the last line (the numbers)
  num_stacks = stack_lines[-1].split.length

  # Parse stacks (bottom-up, excluding the number line)
  stacks = Array.new(num_stacks) { [] }
  stack_lines[0...-1].each do |line|
    num_stacks.times do |i|
      pos = 1 + i * 4  # Position of crate letter
      if pos < line.length && line[pos] != ' '
        stacks[i] << line[pos]
      end
    end
  end

  # Reverse so bottom is at index 0
  stacks.each(&:reverse!)

  # Parse moves
  moves = []
  move_lines.each do |line|
    if line =~ /move (\d+) from (\d+) to (\d+)/
      count = $1.to_i
      from_stack = $2.to_i - 1  # 0-indexed
      to_stack = $3.to_i - 1
      moves << [count, from_stack, to_stack]
    end
  end

  [stacks, moves]
end

def part1(stacks, moves)
  stacks = stacks.map(&:dup)
  moves.each do |count, from_stack, to_stack|
    count.times do
      crate = stacks[from_stack].pop
      stacks[to_stack] << crate
    end
  end
  stacks.map { |stack| stack.last }.join
end

def part2(stacks, moves)
  stacks = stacks.map(&:dup)
  moves.each do |count, from_stack, to_stack|
    # Move multiple crates at once (preserve order)
    crates = stacks[from_stack].pop(count)
    stacks[to_stack].concat(crates)
  end
  stacks.map { |stack| stack.last }.join
end

def main
  script_dir = File.dirname(File.expand_path(__FILE__))
  input_file = File.join(script_dir, '..', 'input.txt')

  stacks, moves = parse_input(input_file)

  puts "Part 1: #{part1(stacks, moves)}"
  puts "Part 2: #{part2(stacks, moves)}"
end

main
